accepted/wasmtime-compile-time-builtins.md

```diff
@@ -0,0 +1,400 @@
+`wasmtime::CodeBuilder::define_compile_time_module_instance` method for
+compile-time core module instance. It is simply excluded from this RFC because
+our current use cases all use component interfaces. Once the above proposal for
+compile-time components is implemented, it should be relatively straightforward
+to reuse that infrastructure for core modules, and we can discuss that
+possibility at that future point in time.
+
+## Winch and compile-time builtins?
+
+There was originally some question around whether it even made sense to allow
+compile-time builtins with Winch, since Winch will not inline calls to these
+compile-time builtins, defeating most of their motivation. However, satisfying a
+Wasm's imports at compile time, making it so that they need not be provided at
+instantiation time, is still valuable for Winch users. Furthermore, nothing laid
+out in the above proposal is fundamentally incompatible with Winch, or relies on
+anything specific to any of our Wasm compilers. The only compiler-specific piece
+is the implementation of each of the Wasmtime intrinsics' stubs, and these are
+isolated and should be relatively straightforward to implement. Therefore, while
+we might skip Winch support in an initial "MVP" implementation, there is no
+reason not to support compile-time builtins with Winch in the fullness of time.
+
+# Open questions
+[open-questions]: #open-questions
+
+* **How do we specify which resource type's table we want to access in the
+  `resource.address` intrinsic?**
+
+  The `resource.address` signature above is not actually valid: the component
+  model's interface types do not provide a way to define a function that takes
+  *any* resource, regardless of where or when it was defined, as an
+  argument. And in fact, if I remember correctly, we use different index spaces
+  for different types of resources, so resource index `r` could be valid in
+  multiple different resource tables.
+
+  This intrinsic kind of wants to be a canonical builtin (like `resource.drop`
+  or `future.new`) rather than a regular function, so we can provide a resource
+  type as an immediate to disambiguate between different types of resources. But
+  defining new canonical builtins is entering the realm of extending the Wasm
+  language, rather than just providing powerful imports to certain components,
+  and I don't think we should go down that route.
+
+  I suppose we could provide the type index of the resource type as a dynamic
+  argument -- in practice it should always be a constant so we can figure out
+  which resource table to access at compile time. But if it is not a constant,
+  what do we even do? Call out to the host? The whole point of this feature is
+  to avoid such calls... Perhaps this isn't so bad, and it just becomes another
+  array indirection: index into the array of resource tables, then index into
+  the array of table elements?
+
+  Anyone have any other ideas?
```

Comment:
you could use LLVM intrinsics as inspiration and have the import's name tell you which type and resources to use, e.g. have `resource.address.t1` for a resource in table `1` or something.